### PR TITLE
Musl native toolchain + all packages necessary for static curl, git builds

### PR DESCRIPTION
**Pull request by satmandu**
- Part of the curl package breakup, to allow building static.
- Please see if `/usr/local/musl/bin/gcc -dumpversion` works.
- Also see if `crew build musl_zlib` works.
- Only tested **in container**.

Works properly:
- [x] x86_64 
- [ ] armv7l
- [ ] i686

### CODE REVIEW

**Comment by satmandu** [comment]
@ThatGeekyWeeb does this work for you in i686?

File: packages/musl_native_toolchain.rb [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    ENV['CREW_FHS_NONCOMPLIANCE_ONLY_ADVISORY'] = '1'
+    ENV['CREW_NOT_STRIP'] = '1'
+    ENV['CREW_SHRINK_ARCHIVE'] = '0'
+    warn_level = $VERBOSE
+    $VERBOSE = nil
+    load "#{CREW_LIB_PATH}lib/const.rb"
+    $VERBOSE = warn_level
+    FileUtils.mkdir_p "#{CREW_DEST_PREFIX}/musl/lib"
+    Dir.chdir("#{CREW_DEST_PREFIX}/musl") do
+      FileUtils.ln_s 'lib', 'lib64'
```
Comment (uberhacker) [inline packages/musl_native_toolchain.rb:45]: This only needs to be done on x86_64, correct?

Comment (satmandu) [reply packages/musl_native_toolchain.rb]: Yup. Added a condition.

**Comment by ThatGeekyWeeb** [comment]
iirc musl does not build for i686

**Comment by ThatGeekyWeeb** [comment]
hmm but you seem to have a package, but I have no way to test it, don't have an i686 chromebook; mine is armv7

**Comment by ThatGeekyWeeb** [comment]
should be noted that it is impossible to build a static curl with ssh support; or rather I tested and attempted to for months with 0 success

**Comment by ThatGeekyWeeb** [comment]
(this also means libcurl will lack ssh support and pkgs such as git that use libcurl will be unable to use ssh)

**Comment by satmandu** [comment]
> should be noted that it is impossible to build a static curl with ssh support; or rather I tested and attempted to for months with 0 success 

I want to take a gander at this. 😂

**Comment by ThatGeekyWeeb** [comment]
> > should be noted that it is impossible to build a static curl with ssh support; or rather I tested and attempted to for months with 0 success
> 
> I want to take a gander at this. joy

I'm p sure it's due to the fact that libssh2 and libssh both just can't be used statically; I tired with both when I did my testing

File: packages/musl_native_toolchain.rb (hunk 1 above)
Comment (satmandu) [reply packages/musl_native_toolchain.rb]: Removed that entirely...

**Comment by satmandu** [comment]
@ThatGeekyWeeb 
```
chronos@cheekon-x86_64 /usr/local/musl/lib $ curl --version
curl 7.78.0 (x86_64-pc-linux-gnu) libcurl/7.78.0 OpenSSL/1.1.1k zlib/1.2.11 brotli/1.0.9 zstd/1.5.0 c-ares/1.17.2 libidn2/2.3.2 libssh/0.9.5/openssl/zlib nghttp2/1.44.0
Release-Date: 2021-07-21
Protocols: dict file ftp ftps gopher gophers http https mqtt pop3 pop3s scp sftp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS brotli GSS-API HSTS HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB SPNEGO SSL TLS-SRP UnixSockets zstd
chronos@cheekon-x86_64 /usr/local/musl/lib $ ldd `which curl`
        not a dynamic executable
```

**Comment by uberhacker** [comment]
On i686, I get:
```
$ crew build musl_zlib
musl_zlib: zlib is a massively spiffy yet delicately unobtrusive compression library.
Downloading source...
######################################################################## 100.0%
Musl_zlib archive downloaded.
Unpacking archive using 'tar', this may take a while...
Building from source, this may take a while...
Rename all *.la files to *.la_tmp
Compiler error reporting is too harsh for ./configure (perhaps remove -Werror).
** ./configure aborting.
musl_zlib failed to build: `PATH=/usr/local/musl/bin:/home/chronos/user/.cargo/bin:/usr/local/pkg/bin:/usr/local/pkg/sbin:/usr/local/bin:/usr/bin:/bin:/opt/bin:/usr/local/libexec/busybox/bin:/usr/local/libexec/busybox/sbin       CC='/usr/local/musl/bin/i686-linux-musl-gcc'       CXX='/usr/local/musl/bin/i686-linux-musl-g++'       LD=/usr/local/musl/bin/i686-linux-musl-ld.gold       PKG_CONFIG_LIBDIR=/usr/local/musl/lib/pkgconfig       CFLAGS='-B/usr/local/musl/include -flto -pipe -O3 -ffat-lto-objects -fipa-pta -fno-semantic-interposition -fdevirtualize-at-ltrans  -fno-stack-protector -fcommon'       CXXFLAGS='-B/usr/local/musl/include -flto -pipe -O3 -ffat-lto-objects -fipa-pta -fno-semantic-interposition -fdevirtualize-at-ltrans  -fno-stack-protector -fcommon'       CPPFLAGS='-I/usr/local/musl/include -fcommon'       LDFLAGS='-L/usr/local/musl/lib -flto -static' ./configure --prefix=/usr/local/musl           --static` exited with 1
```

**Comment by satmandu** [comment]
@uberhacker  Can you post the part of `configure.log` from that build so we can see what was setting that off?

**Comment by ThatGeekyWeeb** [comment]
> @ThatGeekyWeeb
> 
> ```
> chronos@cheekon-x86_64 /usr/local/musl/lib $ curl --version
> curl 7.78.0 (x86_64-pc-linux-gnu) libcurl/7.78.0 OpenSSL/1.1.1k zlib/1.2.11 brotli/1.0.9 zstd/1.5.0 c-ares/1.17.2 libidn2/2.3.2 libssh/0.9.5/openssl/zlib nghttp2/1.44.0
> Release-Date: 2021-07-21
> Protocols: dict file ftp ftps gopher gophers http https mqtt pop3 pop3s scp sftp smb smbs smtp smtps telnet tftp
> Features: alt-svc AsynchDNS brotli GSS-API HSTS HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB SPNEGO SSL TLS-SRP UnixSockets zstd
> chronos@cheekon-x86_64 /usr/local/musl/lib $ ldd `which curl`
>         not a dynamic executable
> ```

I'd have to look into this more

---

But I think it might be smarter to, use some kinda of config variable to set which libc is being used and compile based on that config var instead of build custom musl packges; most packages should work instantly with musl and other that don't will require patches; should be noted that these patches will have no effect on glibc builds

none the less whats being done here seems fine unless you want to look into the above

**Comment by ThatGeekyWeeb** [comment]
That is, unless we are simply adding the musl cross build tool chain and not adding musl in it's entirety; if we are just doing a cross build tool chain addition; maybe this is also a good time to look into the major advantages musl would have over glibc for chromebrew as a whole

**Comment by satmandu** [comment]
> That is, unless we are simply adding the musl cross build tool chain and not adding musl in it's entirety; if we are just doing a cross build tool chain addition; maybe this is also a good time to look into the major advantages musl would have over glibc for chromebrew as a whole

This musl tool chain is really only useful for building static linked apps, I think.

Which is nice, since in theory this bypasses the orphaned Glibc on older EOL machines, but also because it is so much easier to get static builds of apps with musl than with Glibc which are also in theory backwards compatible with older Glibc installations.

Any thoughts on why we might be getting breakage on i686 with this though?

Do we know anyone besides @uberhacker who has i686 machines we can have test these i686 builds?

@uberhacker do you have more than one i686 machine? I just want to rule out hardware failure...

**Comment by ThatGeekyWeeb** [comment]
I would consider musl with i686 a lost cause; void doesn't provide a i686 port of musl; I don't recall if alpine provides one, but if they do; might be smart to dig around in the APORTS repo to figure out how

**Comment by ThatGeekyWeeb** [comment]
I checked alpine and aports; zlib should work fine with i686; there is likely an issue with the toolchain

**Comment by satmandu** [comment]
@ThatGeekyWeeb I think I got the static git & static curl with ssh working. See https://github.com/skycocker/chromebrew/pull/6078 and https://github.com/skycocker/chromebrew/pull/6065

**Pull request merged**